cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -173,6 +181,30 @@ protected void startUp() throws Exception {
     httpService.startAndWait();
     runningThread = Thread.currentThread();
     createLocalDatasets();
+    workflow = initializeWorkflow();
+  }
+
+  private Workflow initializeWorkflow() throws Exception {
+    Class<?> clz = Class.forName(workflowSpec.getClassName(), true, program.getClassLoader());
+    if (!Workflow.class.isAssignableFrom(clz)) {
+      throw new IllegalStateException(String.format("%s is not Workflow.", clz));
+    }
+
+    @SuppressWarnings("unchecked")
+    Class<? extends Workflow> workflowClass = (Class<? extends Workflow>) clz;
+    Workflow workflow = new InstantiatorFactory(false).get(TypeToken.of(workflowClass)).create();
+
+    TransactionContext transactionContext = basicWorkflowContext.getDatasetCache().newTransactionContext();
+    transactionContext.start();
+
+    try {
+      workflow.initialize(basicWorkflowContext);
+    } catch (Throwable t) {
+      LOG.error("Failed to execute the initialize method of the Workflow.", t);
```

Comment:
Yes. We do have access to the `ProgramRunId` instance. Using that in the error message.